app/graphql/types/rule.rb

```diff
@@ -14,37 +14,66 @@ class Rule < Types::BaseObject
     field :severity, String, null: false
     field :remediation_available, Boolean, null: false
     field :profiles, [::Types::Profile], null: true
-    field :identifier, ::Types::RuleIdentifier, null: true
-    field :references, [::Types::RuleReference], null: true,
-                                                 extras: [:lookahead]
+    field :identifier, String, null: true
+    field :references, String, null: true
     field :compliant, Boolean, null: false do
-      argument :system_id, String, 'Is a system compliant?', required: true
+      argument :system_id, String, 'Is a system compliant?',
+               required: false
+      argument :profile_id, String, 'Is a system compliant with this profile?',
+               required: false
     end
 
     def compliant(args = {})
-      return false unless system_id(args) && profile_id(args)
+      system_id(args) &&
+        profile_id(args) &&
+        %w[pass notapplicable notselected].include?(
+          context[:rule_results][object.id][profile_id(args)]
+        )
+    end
 
-      object.compliant?(
-        ::Host.find(system_id(args)),
-        ::Profile.find(profile_id(args))
-      )
+    def references
+      if context[:"rule_references_#{object.id}"].nil?
+        ::CollectionLoader.for(::Rule, :rule_references)
+                          .load(object).then do |references|
+          references.map { |ref| [ref.href, ref.label] }.to_json
+        end
+      else
+        references_from_context
+      end
     end
 
-    def references(lookahead:)
-      selected_columns = lookahead.selections.map(&:name) &
-                         ::RuleReference.column_names.map(&:to_sym)
-      object.references.distinct.reorder('')
-            .pluck(selected_columns).map do |reference|
-        selected_columns.zip([reference].flatten).to_h
+    def references_from_context
+      ::RecordLoader.for(::RuleReference)
+                    .load_many(context[:"rule_references_#{object.id}"])
+                    .then do |references|
+        references.map { |ref| [ref.href, ref.label] }.to_json
```

Comment:
The reason why the references are printed as simple strings is that I could not find a way of rendering them as RuleReference (the GraphQL type) without forcing GQL to find them in the db again. Finding them on the db triggers a SELECT per rule_reference, which is a huge amount when you want to get all references for all rules/profiles in a host (system details page)